Lib/zipfile/__init__.py

```diff
@@ -603,6 +603,26 @@ def from_file(cls, filename, arcname=None, *, strict_timestamps=True):
 
         return zinfo
 
+    @classmethod
+    def for_name(cls, filename, archive, *, date_time=None):
+        """Construct an appropriate ZipInfo from a filename and a ZipFile.
+
+        The *filename* is expected to be the name of a file in the archive.
+
+        If *date_time* is not specified, the current local time is used.
```

Comment:
Currently, this parameter isn't used. If we decide to only expose the function privately, I'd say to remove this unused parameter.

If we expose it publicly, we should consider whether or not we expect users to need this functionality. If users really need this convenience feature, perhaps it should be part of the default constructor and not part of this custom constructor (and then just passed-through here).

In fact, the more I think about it, I'm thinking the interface should be this:

```
def for_archive(cls, archive, *args, **kwargs):
    self = cls(*args, **kwargs)
    self.compress_type = archive.compression
    ...
```

And then add the `date_time=None` support to the default constructor.

Actually, now that I think about it, the wrapping is probably unnecessary. What about adding a helper for the default constructor:

```
class ZipInfo:
    def for_archive(self, archive):
        self.compress_type = archive.compression
        ...
        return self
```

And then callers could compose it with the default constructor:

```
zi = DirtyZipInfo(filename).for_archive(zf)
# or
zinfo = ZipInfo(zinfo_or_arcname).for_archive(self)
```

I think I like that best, because it keeps the concerns separate and avoids wrapping and passthrough, but also enables arbitrary ZipInfo objects to have this behavior.

Reply:
By the way, the reason I added `date_time` is because it was a method I publicly exposed. In particular, I'd assume that it should support the same functionalities as the default constructor (although, in this case the defaults differ; but since we were only using it with an explicit local time, I just went for the simplest choice).

Reply:
It sounded like we were converging on both liking the `for_archive(self, archive) -> Self` approach. Were you planning to revise this PR accordingly?

Reply:
I wasn't planning to do anything until I got a green light. I'll try to see how to revise the PR after some sleep (at first glance, it seems quite straightforward).

Reply:
I'd be just as happy with a distinct PR.

Reply:
So I started something but I'm actually wondering which kind of attributes should be copied. For instance, compression type and level are relevant, but external attributes depend on the filename (at least in `writestr`). 

The composition approach is quite good, assuming that we don't use the object in different contexts. So maybe, we should return a fresh copy instead (a bit like `copy.replace`). I can imagine people writing something like:

```
zi1 = ZipInfo(...)
zi2 = zi1.for_archive(a1)
# oh no, zi1 has been changed
```

and forget about zi1 having been mutated. Because of that, I feel that having `ZipInfo.for_archive` or `ZipInfo.from_info(previous_zipinfo, archive)` could be useful. What do you think?

Sorry for having said "yes" and saying now "I'm not sure".

Reply:
That's a good point. I too was uneasy about both mutating the object and returning self. It's an impure approach and a little dangerous, as you illustrate.

> `ZipInfo.for_archive` or `ZipInfo.from_info(previous_zipinfo, archive)` could be useful.

I don't care for `.from_info` as it makes the primary use-case messy:

```
zi = DirtyZipInfo.from_info(zipfile.ZipInfo(filename), zf)
```

For the same reason, I don't want to have to mutate a zip info but not return it, because it makes the use-case require two lines every time:

```python
zi = DirtyZipInfo(filename)
zi.for_archive(zf)
# use zi
```

I'm thinking we should take one of two options:

## Don't mutate self

As you've proposed, `for_archive()` could create a new instance of `self.__class__` with the same `vars(self)` and return that. This approach is safer, if a little inefficient and likely less elegant in implementation.

It's a little awkward because `self.__class__` might be a subclass of `ZipInfo` (such as with `DirtyZipInfo`), and `ZipInfo` can't be confident about how to construct a subclass, which in general could have additional required constructor parameters or instance variables:

```python
class OtherZipInfo(zipfile.ZipInfo):
    def __init__(self, special_param):
        self.special_param = special_param
        super().__init__(...)
```

In either case (unique vars() or args), I guess that subclass would also be required to provide their own `for_archive`.

## Mutate self, return self

Accept the less safe approach and just be sure to document the behavior, e.g.:

```python
def for_archive(self, archive: ZipFile) -> Self:
    """
    Configure attributes of this ZipInfo based on the archive and other defaults.

    Mutates the object in place but also returns itself for chained operations.
    """
    ...
```

This option would be nicer if there were a well-established pattern for such methods, something like [jaraco.functools.chainable](https://jaracofunctools.readthedocs.io/en/latest/#jaraco.functools.chainable).

```python
@chainable
def for_archive(self, archive: ZipFile) -> None:
    self.compression = archive.compression
    ...
```

Such a `chainable` decorator would be a more readily and consistently understood signal that the method always returns a reference to itself. Alas, no such decorator exists in the stdlib.


----

Overall, I'm leaning toward "mutate self, return self," because it provides the nicest ergonomics and because the weaknesses can be easily overcome with documentation.

WDYT?

Reply:
Err... I personally prefer not mutating self but since I don't use ZipInfo much I think a good documentation would be sufficient. 

What I hit when I was writing the interface is that the filename for the zipinfo in `writestr` is not the filename of the archive itself. Since the external attributes depend on that filename in particular, we would simply save the compression and compress level update but not the external attributes part:

```py
def writestr(self, zipinfo_or_arcname, ...):
	zinfo = ZipInfo(zipinfo_or_arcname, date_time=time.localtime(time.time())[:6])
	zinfo = zinfo.for_archive(self)
	if zinfo.filename.startswith('/'):
		zinfo.external_attrs = ...  # I don't remember the exact attribute name
	else:
		zinfo.external_attrs = ...
``` 

I think ZipInfo could have an interface for actually correctly deriving the default external attributes depending on the filename (or we can make a private helper for that):

```py
def get_default_external_attributes(filename: str) -> int: ...
```

What do you think?

Reply:
> What I hit when I was writing the interface is that the filename for the zipinfo in `writestr` is not the filename of the archive itself. Since the external attributes depend on that filename in particular, we would simply save the compression and compress level update but not the external attributes part:

It'll be insufficient to only extract half of the operations. The goal is to eliminate `test_path.DirtyZipInfo.for_name`.

I don't see how it's the case that

> the external attributes depend on [the filename of the archive itself] in particular

In `ZipFile.writestr`, external attributes are calculated against `zinfo.filename` (the filename of the ZipInfo object and not the filename of the archive itself). When moving that functionality from `ZipFile` to `ZipInfo`, it changes the reference from `zinfo.filename` to `self.filename`, as is drafted in the current PR.

What issue are you encountering relating to external attrs? Can you push the code somewhere that demonstrates the issue?

Reply:
Urgh, I should stop reviewing at 2 AM. The thing is `archive` also has a `filename` so I thought about updating `zinfo.filename = archive.filename` but that's wrong. Ok my interrogations are resolved (sorry).

Reply:
In order to unblock things, in 44e7b75ffe6, I've implemented what I believe is very close to where we were heading, and we can revisit in subsequent issues/PRs if needed.